projects/imagenet/experiments/base.py

```diff
@@ -25,10 +30,14 @@
 import torch
 
 import nupic.research.frameworks.pytorch.models.resnets
+from nupic.research.frameworks.pytorch.imagenet import ImagenetExperiment, mixins
+
+
+class RezeroedKWinnersImagenetExperiment(mixins.RezeroWeights,
```

Comment:
I'm a bit worried about order of operations. Some mixins call super first thing, others call them at other points. This makes it more flexible but could be hard to debug. As an aid to debugging, is there an easy way to log the exact sequence of mixin methods that are called?  In addition, I think docstrings for the mixins would be important. Perhaps we should have a convention to note any issues in the docstring (e.g. this mixin assumes it will be called before x or after y).

Reply:
Ok, to address this I added a `self.execution_order` to the ImagenetExperiment. Each mixin and subclass documents itself during `__init__` by updating the execution order. During `setup_experiment`, this execution order is printed.

So, for example, when I include the Profile mixin, the printed dict will include:

```
'run_epoch': ['Profile begin', 'ImagenetExperiment.run_epoch', 'Profile end'],
```

Reply:
Nice!

Reply:
Does this guarantee the exact execution order though? If a mixin calls super first, then does its thing, vs calls super after it does its thing the above list would still be the same but the true "algorithmic execution order" is reversed.

Reply:
- If the mixin calls super at the beginning, it appends to the execution order.
- If the mixin calls super at the end, it prepends to the execution order

This doesn't happen magically; the mixin's __init__ method carefully chooses between appending/prepending/both.

Reply:
It is simpler to just do `help(RezeroedKWinnersImagenetExperiment)` or `Experiment.mro()` to display the method resolution order instead of managing ourselves

Reply:
But I don't think that would address Subutai's original point? Sometimes super() is called at the beginning of a method, sometimes at the end, and sometimes it is sandwiched between calls (like with profiling). This approach documents all of that.

Reply:
> * If the mixin calls super at the beginning, it appends to the execution order.
> * If the mixin calls super at the end, it prepends to the execution order
> 
> This doesn't happen magically; the mixin's `__init__` method carefully chooses between appending/prepending/both.

Ah ok, got it. Makes sense.